crates/wasmprinter/src/operator.rs

```diff
@@ -1,36 +1,81 @@
-use super::{Print, Printer, State};
+use super::{Config, Print, PrintTermcolor, Printer, State};
 use anyhow::{anyhow, bail, Result};
-use wasmparser::{BlockType, BrTable, Catch, MemArg, Ordering, RefType, TryTable, VisitOperator};
+use termcolor::{Ansi, NoColor};
+use wasmparser::{
+    operator_arity, BlockType, BrTable, Catch, CompositeInnerType, CompositeType, FrameKind,
+    MemArg, ModuleArity, Ordering, RefType, SubType, TryTable, VisitOperator,
+};
 
-pub struct PrintOperator<'printer, 'state, 'a, 'b> {
-    pub(super) printer: &'printer mut Printer<'a, 'b>,
-    pub(super) op_offset: usize,
+pub struct OperatorState {
+    op_offset: usize,
     nesting_start: u32,
-    state: &'state mut State,
     label: u32,
     label_indices: Vec<u32>,
     sep: OperatorSeparator,
 }
 
+impl OperatorState {
+    pub fn new(printer: &Printer, sep: OperatorSeparator) -> Self {
+        OperatorState {
+            op_offset: 0,
+            nesting_start: printer.nesting,
+            label: 0,
+            label_indices: Vec::new(),
+            sep,
+        }
+    }
+}
+
+pub struct PrintOperator<'printer, 'state, 'a, 'b> {
+    pub(super) printer: &'printer mut Printer<'a, 'b>,
+    state: &'state mut State,
+    operator_state: &'printer mut OperatorState,
```

Comment:
If you'd like I think it'd make sense to store this as `OperatorState` instead of `&mut OperatorState` since the state itself won't outlive the print operation (unlike `State` which does outlive). Having a pointer here works fine though, so mostly a stylistic thing.

Reply:
It's a little tricky because the `PrintOperatorFolded` also constructs a temporary `PrintOperator` in its visitor function and wants to let it borrow (but not consume) its `OperatorState`.

Reply:
Ah ok it's fine to leave as-is in that case 👍